### ISSUE TITLE
i8xN target types

### ISSUE THREAD

**Issue by Wunkolo** [ispc/ispc#1607]
Currently, the list of target types only seems to include one `i8` target for the x86-ISA SIMD extensions(`sse4-i8x16`).

```
       host, sse2-i32x4, 
       sse2-i32x8, sse4-i32x4, sse4-i32x8, sse4-i16x8, sse4-i8x16, avx1-i32x4, 
       avx1-i32x8, avx1-i32x16, avx1-i64x4, avx2-i32x4, avx2-i32x8, 
       avx2-i32x16, avx2-i64x4, avx512knl-i32x16, avx512skx-i32x16, 
       avx512skx-i32x8, generic-x1, generic-x4, generic-x8, generic-x16, 
       generic-x32, generic-x64, *-generic-x16, neon-i8x16, neon-i16x8, 
       neon-i32x4, neon-i32x8
```
Are there any plans or blockers stopping support for targets such as `avx512skx-i8x64` or `avx2-i8x32`? These would be especially useful when making kernels that deal with text and image data.

Right now (ispc 1.12.0) returns the following assembly with `-O3 --target=avx512skx-i32x16`:

```c
int8 func(int8 a, int8 b)
{
    return a + b;
}
```

```
func___vytvyt:                          # @func___vytvyt
        vpaddb  xmm0, xmm1, xmm0
        ret
```

It is only using the `xmm` registers, adding x16 at a time, when it could be adding x64 at a time and doing the equivalent of `_mm512_add_epi8` with the `zmm` registers.

I could possibly be misunderstanding these flags though and `avx512skx-i8x64` means it will try and force a 32-bit addition to use 8-bit additions which would involve a _lot_ more code moving around than usual.
How would I try and get ispc to emit the more optimal 8-bit arithmetic though?

**Comment by dbabokin** [comment]
Indeed, for x86 we have only sse4 target, which supports i8 as base type. The reason why we haven't implemented avx2 and avx512 targets with i8 base type is basically lack of demand for it. So if someone needs these target, these targets can be implemented. Actually it's not much work to add these targets. Though tuning them might be more work.

If someone is willing to contribute these targets, I would support that. Otherwise would be good to see the project that is going to use them - just to see that the work is not wasted and is actually used.

**Comment by Wunkolo** [comment]
Thank you for the reply.
I would imagine it would be very useful for ASCII text processing and encryption and base64 encoding and image processing and such of the sort.
I could contribute such a thing myself if there were some kind of contribution guidelines or a way to have a high-level understanding of the codebase to implement it myself and make a PR of some sort.

Not sure if this is worthy of another issue, but it would also be very useful to have character-literal support in ISPC as well to facilitate its application in ascii-text processing. When I tried to scope an acceleration of base85 and base64 encoding in godbolt using ispc:
I tried to use a `'!'` (33) and `'u'`(117) literal only to find that ispc does not support it when they are [basically the same as an `int`](https://en.cppreference.com/w/c/language/character_constant).

**Comment by dbabokin** [comment]
It took more efforts to implement `avx512skx-i8x64` and `avx512skx-i16x32` targets that I've expected. I've just completed these targets. They are functionally correct, but stdlib is not well optimized yet. The changes require LLVM 10.0 with set of patches (included in my branch). 9.0 is not supported. LLVM trunk is known to have regressions at this moment. If you prefer to play with the changes before they are committed, here's my branch: https://github.com/dbabokin/ispc/tree/avx512_i8_i16

**Comment by dbabokin** [comment]
I've added avx2 (#1746) and avx512skx (#1744) versions of i8 and i16 targets: ``avx2-i8x32``, ``avx2-i16x16``, ``avx512skx-i8x64``, and ``avx512skx-i16x32`` to trunk. They are not fully optimized, but fully functional (all tests are passing). They are available in the downloadable trunk builds for Windows and Linux. For build from source, avx512skx versions require patched 10.0 LLVM version (patches are needed for specifically for these targets, so no way around them). avx2 targets are available with any supported LLVM version (8.0 and later). Everyone is welcome to try and give feedback!

**Comment by dbabokin** [comment]
I assume this should be considered implemented and should be closed.
